packages/flutter/lib/src/material/page_transitions_theme.dart

```diff
@@ -672,13 +748,251 @@ class ZoomPageTransitionsBuilder extends PageTransitionsBuilder {
     return _ZoomPageTransition(
       animation: animation,
       secondaryAnimation: secondaryAnimation,
-      allowSnapshotting: allowSnapshotting && (route?.allowSnapshotting ?? true),
+      allowSnapshotting: allowSnapshotting && route.allowSnapshotting,
       allowEnterRouteSnapshotting: allowEnterRouteSnapshotting,
       child: child,
     );
   }
 }
 
+typedef AndroidBackGestureTransitionBuilder<T> = Widget Function(
+    PageRoute<T> route,
+    BuildContext context,
+    Animation<double> animation,
+    Animation<double> secondaryAnimation,
+    Widget child,
+    AndroidBackEvent? startBackEvent,
+    AndroidBackEvent? currentBackEvent);
+
+class AndroidBackGesturePageTransitionsBuilder extends PageTransitionsBuilder {
+  const AndroidBackGesturePageTransitionsBuilder({
+    this.parent = const ZoomPageTransitionsBuilder(),
+    this.backGestureTransitionBuilder = defaultBackGestureTransitionBuilder,
+  });
+
+  final PageTransitionsBuilder parent;
+  final AndroidBackGestureTransitionBuilder<dynamic>
+      backGestureTransitionBuilder;
+
+  @override
+  Widget buildTransitions<T>(
+    PageRoute<T> route,
+    BuildContext context,
+    Animation<double> animation,
+    Animation<double> secondaryAnimation,
+    Widget child,
+  ) {
+    return AndroidBackGestureDetector(
+      controller: route.controller!, // protected access
+      navigator: route.navigator!,
+      enabledCallback: () =>
+          CupertinoRouteTransitionMixin.isPopGestureEnabled(route),
+      builder: (BuildContext context, AndroidBackEvent? startBackEvent,
+          AndroidBackEvent? currentBackEvent) {
+        final bool linearTransition =
+            CupertinoRouteTransitionMixin.isPopGestureInProgress(route);
+
+        if (linearTransition) {
+          return backGestureTransitionBuilder(
+            route,
+            context,
+            animation,
+            secondaryAnimation,
+            child,
+            startBackEvent,
+            currentBackEvent,
+          );
+        }
+
+        return parent.buildTransitions(
+          route,
+          context,
+          animation,
+          secondaryAnimation,
+          child,
+        );
+      },
+    );
+  }
+
+  static Widget defaultBackGestureTransitionBuilder<T>(
+    PageRoute<T> route,
+    BuildContext context,
+    Animation<double> animation,
+    Animation<double> secondaryAnimation,
+    Widget child,
+    AndroidBackEvent? startBackEvent,
+    AndroidBackEvent? currentBackEvent,
+  ) {
+    return AndroidBackGestureTransition(
+      animation: animation,
+      secondaryAnimation: secondaryAnimation,
+      startBackEvent: startBackEvent,
+      currentBackEvent: currentBackEvent,
+      child: child,
+    );
+  }
+}
+
+typedef AndroidBackGestureDetectorWidgetBuilder = Widget Function(
+    BuildContext context,
+    AndroidBackEvent? startBackEvent,
+    AndroidBackEvent? currentBackEvent);
+
+class AndroidBackGestureDetector extends StatefulWidget {
```

Comment:
What if you split out just a simple "gesture detector" widget with callback parameters startBackGesture, updateBackGestureProgress, commitBackGesture, and cancelBackGesture? I feel like that might be nice to work with, even if it's just private.